### ISSUE TITLE
s6 v3 and supervisor do not like when add-on protection mode is disabled

### ISSUE THREAD

**Issue by loganmarchione** [home-assistant/supervisor#3642]
### Describe the issue you are experiencing

# TL;DR

I develop an add-on called [prometheus_node_exporter](https://github.com/loganmarchione/hassos-addons/tree/main/prometheus_node_exporter). I am by no means a professional developer, but was told I may have stumbled upon a bug with the way the new s6 v3 and supervisor handle protection mode being disabled. I was [asked to open an issue](https://community.home-assistant.io/t/hello-world-example-addon-from-developer-docs-stopped-working-s6-overlay-issue/421486/25). See below for steps to re-create.

----

# Long version

This is my first bug report, let me know if you need more info.

**Background**:
* Prometheus [server](https://github.com/prometheus/prometheus) (not my add-on) is a time-series metrics database (similar to influxd).
* Prometheus also creates a client called [node_exporter](https://github.com/prometheus/node_exporter) that exports node data (CPU, memory, disk, etc...) to a Prometheus server.
* My add-on is a Home Assistant version of that node_exporter client. I use it to monitor my Home Assistant virtual machine metrics and export that data to my separate prometheus server instance.
* Side note: If you could integrate node metrics into the [official Prometheus integration](https://www.home-assistant.io/integrations/prometheus/), that would be great and would negate the need for my add-on completely 😅.

**Problem**:
* Like the [glances add-on](https://github.com/hassio-addons/addon-glances), prometheus_node_exporter requires access to host data that is outside of the container. As such, protection mode must be disabled to gather all metrics. It will technically work without protection mode, but it's missing metrics.
* Since the recent [s6 v3 update](https://developers.home-assistant.io/blog/2022/05/12/s6-overlay-base-images), new installs of my add-on have been [reported broken](https://github.com/loganmarchione/hassos-addons/issues/2).
* I took the steps in the blog post, including making my scripts executable and updating my `finish` script. I had already set `init: false`, so that didn't change. However, the add-on still reported `s6-overlay-suexec: fatal: can only run as pid 1`.
* Working with @mdegat01 on the forums, he pointed out that the add-on does appear to try to start when protection mode is enabled. This would point to a bug in the way the new s6 v3 and supervisor handle protection mode being disabled, right?
* I'm guessing glances isn't affected (yet) because it publishes [images](https://github.com/orgs/hassio-addons/packages?repo_name=addon-glances) to GHCR and was last updated in January, whereas my add-on builds the container on each install, so it immediately broke on the s6 v3 update. I should add that this add-on worked fine before the s6 v3 update.

### What is the used version of the Supervisor?

supervisor-2022.05.2

### What type of installation are you running?

Home Assistant OS

### Which operating system are you running on?

Home Assistant Operating System

### What is the version of your installed operating system?

Home Assistant OS 8.0

### What version of Home Assistant Core is installed?

core-2022.5.4

### Steps to reproduce the issue

1. Clone my repo and checkout the s6_v3 branch
```
git clone https://github.com/loganmarchione/hassos-addons.git
cd hassos-addons
git checkout s6_v3
```
2. Copy/paste the `prometheus_node_exporter` directory into the `hassos-addons` folder of your HA instance
3. Install the add-on, **disable protection mode** (this is needed because the add-on requires access to host data), and start it
4. You will see this error and the add-on will crash
```
s6-overlay-suexec: fatal: can only run as pid 1
```
5. Repeat step 3, but **leave protection mode enabled**, and you'll see a different error.
```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
cont-init: info: running /etc/cont-init.d/node_exporter.sh
[21:37:26] FATAL: PROTECTION MODE IS ENABLED!
[21:37:26] FATAL: 
[21:37:26] FATAL: To be able to use this add-on, you'll need to disable
[21:37:26] FATAL: protection mode on this add-on. Without it, the add-on
[21:37:26] FATAL: is unable to access Docker.
[21:37:26] FATAL: 
[21:37:26] FATAL: Steps:
[21:37:26] FATAL:  - Go to the Supervisor Panel.
[21:37:26] FATAL:  - Click on this add-on.
[21:37:26] FATAL:  - Set the 'Protection mode' switch to off.
[21:37:26] FATAL:  - Restart the add-on.
[21:37:26] FATAL: 
[21:37:26] FATAL: Access to Docker allows you to do really powerful things
[21:37:26] FATAL: including complete destruction of your system.
[21:37:26] FATAL: Please, be sure you know what you are doing before
[21:37:26] FATAL: enabling this feature (and this add-on)!
[21:37:26] FATAL: 
cont-init: info: /etc/cont-init.d/node_exporter.sh exited 1
cont-init: warning: some scripts exited nonzero
s6-rc: warning: unable to start service legacy-cont-init: command exited 1
/run/s6/basedir/scripts/rc.init: warning: s6-rc failed to properly bring all the services up! Check your logs (in /run/uncaught-logs/current if you have in-container logging) for more information.
/run/s6/basedir/scripts/rc.init: fatal: stopping the container.
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped
```
6. This would point to a bug in the way the new s6 v3 and supervisor handle protection mode being disabled, right?
7. Furthermore, if you stop the add-on, comment out [this line](https://github.com/loganmarchione/hassos-addons/blob/8e84bec82cc06be3ae57bade5d2454bead5cff62/prometheus_node_exporter/rootfs/etc/cont-init.d/node_exporter.sh#L6) to remove the protection mode requirement, save, and re-install the add-on, you'll see it starts. This also confirms that something is happening with s6 v3 and protection mode, right?
```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
cont-init: info: running /etc/cont-init.d/node_exporter.sh
[21:42:11] INFO: Checking configuration options...
[21:42:11] WARNING: HTTP Basic Auth is disabled!
cont-init: info: /etc/cont-init.d/node_exporter.sh exited 0
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
services-up: info: copying legacy longrun node_exporter (no readiness notification)
s6-rc: info: service legacy-services successfully started
[21:42:11] INFO: Starting prometheus node exporter...
ts=2022-05-17T01:42:11.995Z caller=node_exporter.go:182 level=info msg="Starting node_exporter" version="(version=1.3.1, branch=HEAD, revision=a2321e7b940ddcff26873612bccdf7cd4c42b6b6)"
ts=2022-05-17T01:42:11.995Z caller=node_exporter.go:183 level=info msg="Build context" build_context="(go=go1.17.3, user=root@243aafa5525c, date=20211205-11:09:49)"
ts=2022-05-17T01:42:11.995Z caller=filesystem_common.go:111 level=info collector=filesystem msg="Parsed flag --collector.filesystem.mount-points-exclude" flag=^/(dev|proc|run/credentials/.+|sys|var/lib/docker/.+)($|/)
ts=2022-05-17T01:42:11.996Z caller=filesystem_common.go:113 level=info collector=filesystem msg="Parsed flag --collector.filesystem.fs-types-exclude" flag=^(autofs|binfmt_misc|bpf|cgroup2?|configfs|debugfs|devpts|devtmpfs|fusectl|hugetlbfs|iso9660|mqueue|nsfs|overlay|proc|procfs|pstore|rpc_pipefs|securityfs|selinuxfs|squashfs|sysfs|tracefs)$
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:108 level=info msg="Enabled collectors"
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=arp
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=bcache
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=bonding
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=btrfs
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=conntrack
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=cpu
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=cpufreq
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=diskstats
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=dmi
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=edac
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=entropy
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=fibrechannel
ts=2022-05-17T01:42:11.996Z caller=node_exporter.go:115 level=info collector=filefd
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=filesystem
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=hwmon
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=infiniband
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=ipvs
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=loadavg
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=mdadm
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=meminfo
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=netclass
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=netdev
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=netstat
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=nfs
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=nfsd
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=nvme
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=os
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=powersupplyclass
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=pressure
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=rapl
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=schedstat
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=sockstat
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=softnet
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=stat
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=tapestats
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=textfile
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=thermal_zone
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=time
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=timex
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=udp_queues
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=uname
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=vmstat
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=xfs
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:115 level=info collector=zfs
ts=2022-05-17T01:42:11.997Z caller=node_exporter.go:199 level=info msg="Listening on" address=:9100
ts=2022-05-17T01:42:11.997Z caller=tls_config.go:231 level=info msg="TLS is disabled." http2=false
```

### Anything in the Supervisor logs that might be useful for us?

```txt
N/A, see above.
```

**Comment by mdegat01** [comment]
@loganmarchione Ah I got it. So its not specifically a protection mode problem, it's actually `host_pid`. In your addon you have [`host_pid: true`](https://github.com/loganmarchione/hassos-addons/blob/main/prometheus_node_exporter/config.json#L25). This setting only takes effect when protection mode is off as you can see here:
https://github.com/home-assistant/supervisor/blob/2c09021427416b84e4f1ad5a12d918fb5e95f9c5/supervisor/docker/addon.py#L253-L257

So that's the actual reason why it broke when protection mode is on and "worked" with it off. It was really because that was turning on and off `host_pid` mode.

That being said I have bad news here, `host_pid` literally won't work with s6 v3. What's actually happening here is that s6 is checking that its init process is PID 1. That will never happen when `host_pid` is `true` because its PID is in the host namespace, that's what the option does.

So we need to update the blog post for this but this is a bit of a problem. It would seem to me you have a few options here:

1. Try removing that. Does your addon still work? If yes, great! If not then next option
2. Switch to using the community addons base image, you can find that [here](https://github.com/hassio-addons/addon-base). This hasn't been updated to s6 v3 yet so it will still work. Basically just add a `build.yaml` and put this in there:
https://github.com/hassio-addons/addon-glances/blob/e2d0d781f5aa62536a9eca7d86a4e3e62c0ccb83/glances/build.yaml#L2-L7

If you need to do #2 then that isn't a long-term solution. Eventually those base images will update too. I would recommend keeping an eye on the Glances addon because this addon has the same situation:
https://github.com/hassio-addons/addon-glances/blob/e2d0d781f5aa62536a9eca7d86a4e3e62c0ccb83/glances/config.yaml#L35

I don't know how we will solve this yet but something will have to be changed as that addon can't use s6 v3 yet either.
